loguru/_locks_machinery.py

```diff
@@ -2,13 +2,46 @@
 import threading
 import weakref
 
+
+class HandlerLockNotReentrant(RuntimeError):
+    pass
+
+
+class _HandlerLockWrapper:
+    """Lock wrapper for the handler.
+
+    This avoids thread-local deadlock by propagating an exception rather than waiting forever.
+    See: https://github.com/Delgan/loguru/issues/712
+    """
+
+    def __init__(self, lock):
+        self._lock = lock
+        self._local = threading.local()
+        self._local.acquired = False
+
+    def __enter__(self):
+        if getattr(self._local, "acquired", False):
+            raise HandlerLockNotReentrant(
```

Comment:
Given my other comment about not handling `HandlerLockNotReentrant` specifically, I think we can just raise a `RuntimeError` here.

Also, given that you suggested to protect all methods, I would make the message a bit more generic (because the error won't necessarily be raised when we "tried to emit a log"). For example:

```python
raise RuntimeError(
    "Could not acquire internal lock because it was already in use (deadlock avoided). "
    "This likely happened because the logger was re-used inside a sink, a signal handler "
    "or a '__del__' method. This is not permitted because the logger and its handlers are "
    "not re-entrant."
)
```

Reply:
👍